src/call/mod.rs

```diff
@@ -616,59 +616,87 @@ impl WriteFlags {
     }
 
     /// Get whether buffer hint is enabled.
-    pub fn get_buffer_hint(self) -> bool {
+    pub fn get_buffer_hint(&self) -> bool {
         (self.flags & grpc_sys::GRPC_WRITE_BUFFER_HINT) != 0
     }
 
     /// Get whether compression is disabled.
-    pub fn get_force_no_compress(self) -> bool {
+    pub fn get_force_no_compress(&self) -> bool {
         (self.flags & grpc_sys::GRPC_WRITE_NO_COMPRESS) != 0
     }
 }
 
 /// A helper struct for constructing Sink object for batch requests.
+///
+/// By default the sink works in normal mode, that is, `start_send` always starts to send message
+/// immediately. But when the `enhance_buffer_strategy` is enabled, the stream will be batched
+/// together as much as possible. The specific rule is listed below:
+/// Set the `buffer_hint` of the non-end message in the stream to true. And set the `buffer_hint`
+/// of the last message to false in `poll_flush` only when there is at least one message with the
+/// `buffer_hint` false, so that the previously bufferd messages will be sent out.
 struct SinkBase {
+    // Batch job to be executed in `poll_ready`.
     batch_f: Option<BatchFuture>,
-    buf: Vec<u8>,
     send_metadata: bool,
+    // Flag to indicate if enhance batch strategy. This behavior will modify the `buffer_hint` to batch
+    // messages as much as possible.
+    enhance_buffer_strategy: bool,
+    // Buffer used to store the data to be sent, send out the last data in this round of `start_send`.
+    // Note: only used in enhanced buffer strategy.
+    buffer: Vec<u8>,
+    // Write flags used to control the data to be sent in `buffer`.
+    // Note: only used in enhanced buffer strategy.
+    buf_flags: Option<WriteFlags>,
+    // Used to records whether a message in which `buffer_hint` is false exists.
+    // Note: only used in enhanced buffer strategy.
+    last_buf_hint: bool,
 }
 
 impl SinkBase {
     fn new(send_metadata: bool) -> SinkBase {
         SinkBase {
             batch_f: None,
-            buf: Vec::new(),
+            buffer: Vec::new(),
+            buf_flags: None,
+            last_buf_hint: true,
             send_metadata,
+            enhance_buffer_strategy: false,
         }
     }
 
     fn start_send<T, C: ShareCallHolder>(
         &mut self,
         call: &mut C,
         t: &T,
-        mut flags: WriteFlags,
+        flags: WriteFlags,
         ser: SerializeFn<T>,
     ) -> Result<()> {
-        // `start_send` is supposed to be called after `poll_ready` returns ready.
-        assert!(self.batch_f.is_none());
```

Comment:
Why removing the assert?

Reply:
I think the assert should be placed as close as possible to the function sending the message. So move it into `start_send_buffer_message`